File: ios-control/src/serverUrlPolicy.ts
const HTTPS_DEFAULT_PORT = 443;

export const DEFAULT_BRIDGE_ORIGIN = 'https://xiaoli.136-110-78-96.sslip.io';

export interface ServerUrlPolicy {
  readonly host: string;
  readonly effectivePort: number;
  readonly origin: string;
  readonly dashboardUrl: string;
  isAllowedNavigation(candidate: string): boolean;
}

function invalidAddress(): Error {
  return new Error('A valid HTTPS Bridge address is required');
}

function parseHttpsUrl(input: string): URL {
  const trimmed = input?.trim();
  if (!trimmed || !/^https:\/\//i.test(trimmed)) {
    throw invalidAddress();
  }

  const authority = trimmed.slice(trimmed.indexOf('//') + 2).split(/[/?#]/, 1)[0];
  if (
    !authority ||
    authority.endsWith(':') ||
    authority.includes('@') ||
    /[\\\u0000-\u001f\u007f]/.test(trimmed)
  ) {
    throw invalidAddress();
  }

  let url: URL;
  try {
    url = new URL(trimmed);
  } catch {
    throw invalidAddress();
  }

  const port = url.port ? Number(url.port) : HTTPS_DEFAULT_PORT;
  if (
    url.protocol !== 'https:' ||
    !url.hostname ||
    url.username !== '' ||
    url.password !== '' ||
    !Number.isInteger(port) ||
    port < 1 ||
    port > 65535 ||
    url.origin === 'null'
  ) {
    throw invalidAddress();
  }

  return url;
}

export function createServerUrlPolicy(input: string): ServerUrlPolicy {
  const url = parseHttpsUrl(input);
  const host = url.hostname.toLowerCase();
  const effectivePort = url.port ? Number(url.port) : HTTPS_DEFAULT_PORT;
  const origin = `https://${host}${effectivePort === HTTPS_DEFAULT_PORT ? '' : `:${effectivePort}`}`;

  return Object.freeze({
    host,
    effectivePort,
    origin,
    dashboardUrl: `${origin}/mobile/`,
    isAllowedNavigation(candidate: string): boolean {
      try {
        const candidateUrl = parseHttpsUrl(candidate);
        const candidatePort = candidateUrl.port ? Number(candidateUrl.port) : HTTPS_DEFAULT_PORT;
        return candidateUrl.hostname.toLowerCase() === host && candidatePort === effectivePort;
      } catch {
        return false;
      }
    },
  });
}

export function isApkDownload(candidate: string): boolean {
  try {
    return /\.apk$/i.test(new URL(candidate).pathname);
  } catch {
    return false;
  }
}

File: android-control/app/src/main/java/cn/xiaoli/control/ServerUrlPolicy.java
package cn.xiaoli.control;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Locale;

/** Defines the one HTTPS origin that the WebView may navigate within. */
public final class ServerUrlPolicy {
    private static final int HTTPS_DEFAULT_PORT = 443;

    private final String host;
    private final int effectivePort;
    private final String origin;

    private ServerUrlPolicy(String host, int effectivePort, String origin) {
        this.host = host;
        this.effectivePort = effectivePort;
        this.origin = origin;
    }

    public static ServerUrlPolicy fromUserInput(String input) {
        URI uri = parseUri(input);
        if (!isHttpsHierarchicalUrl(uri) || uri.getRawUserInfo() != null) {
            throw invalidAddress();
        }

        String parsedHost = uri.getHost();
        int parsedPort = uri.getPort();
        if (parsedHost == null || parsedHost.trim().isEmpty() || !hasValidPort(uri)) {
            throw invalidAddress();
        }

        String canonicalHost = parsedHost.toLowerCase(Locale.ROOT);
        int canonicalPort = parsedPort == -1 ? HTTPS_DEFAULT_PORT : parsedPort;
        int renderedPort = canonicalPort == HTTPS_DEFAULT_PORT ? -1 : canonicalPort;
        try {
            String canonicalOrigin = new URI(
                    "https", null, canonicalHost, renderedPort, null, null, null)
                    .toASCIIString();
            return new ServerUrlPolicy(canonicalHost, canonicalPort, canonicalOrigin);
        } catch (URISyntaxException error) {
            throw invalidAddress();
        }
    }

    public String origin() {
        return origin;
    }

    public String dashboardUrl() {
        return origin + "/mobile/";
    }

    public boolean isAllowedNavigation(String candidate) {
        URI uri;
        try {
            uri = parseUri(candidate);
        } catch (IllegalArgumentException error) {
            return false;
        }

        if (!isHttpsHierarchicalUrl(uri)
                || uri.getRawUserInfo() != null
                || uri.getHost() == null
                || !hasValidPort(uri)) {
            return false;
        }

        int candidatePort = uri.getPort() == -1 ? HTTPS_DEFAULT_PORT : uri.getPort();
        return host.equalsIgnoreCase(uri.getHost()) && effectivePort == candidatePort;
    }

    private static URI parseUri(String input) {
        if (input == null || input.trim().isEmpty()) {
            throw invalidAddress();
        }
        try {
            return new URI(input.trim());
        } catch (URISyntaxException error) {
            throw invalidAddress();
        }
    }

    private static boolean isHttpsHierarchicalUrl(URI uri) {
        return !uri.isOpaque() && "https".equalsIgnoreCase(uri.getScheme());
    }

    private static boolean hasValidPort(URI uri) {
        int port = uri.getPort();
        String authority = uri.getRawAuthority();
        return authority != null
                && !authority.endsWith(":")
                && port != 0
                && port <= 65535;
    }

    private static IllegalArgumentException invalidAddress() {
        return new IllegalArgumentException("A valid HTTPS Bridge address is required");
    }
}
